lib/assert.js

```diff
@@ -269,9 +269,7 @@ function _deepEqual(actual, expected, strict, memos) {
 
   const actualPosition = memos.actual.map.get(actual);
   if (actualPosition !== undefined) {
-    if (actualPosition === memos.expected.map.get(expected)) {
-      return true;
-    }
+    return actualPosition === memos.expected.map.get(expected);
```

Comment:
I think the failed test case is caused by `memos.expected.map.get(expected)` returning `undefined`? To make sure the two are visited in the same order, we need to make sure `expected` has actually been visited before.